spicy/toolchain/include/ast/types/unit-items/field.h

```diff
@@ -29,41 +29,44 @@ static inline Type adaptType(Type type, const std::optional<Expression>& repeat)
 /** AST node for a unit field. */
 class Field : public hilti::NodeBase, public spicy::trait::isUnitItem {
 public:
-    Field(const std::optional<ID>& id, Type type, Engine e, const std::vector<Expression>& args,
+    Field(const std::optional<ID>& id, uint64_t index, Type type, Engine e, const std::vector<Expression>& args,
           std::optional<Expression> repeat, const std::vector<Expression>& sinks,
           std::optional<AttributeSet> attrs = {}, std::optional<Expression> cond = {}, std::vector<Hook> hooks = {},
           Meta m = Meta())
         : NodeBase(nodes((id ? id : _uniquer.get("anon")), detail::adaptType(std::move(type), repeat), node::none,
                          repeat, std::move(attrs), std::move(cond), args, sinks, std::move(hooks)),
                    std::move(m)),
+          _index(index),
           _is_anonynmous(! id.has_value()),
           _engine(e),
           _args_start(6),
           _args_end(_args_start + static_cast<int>(args.size())),
           _sinks_start(_args_end),
           _sinks_end(_sinks_start + static_cast<int>(sinks.size())) {}
 
-    Field(const std::optional<ID>& id, Ctor ctor, Engine e, const std::vector<Expression>& args,
+    Field(const std::optional<ID>& id, uint64_t index, Ctor ctor, Engine e, const std::vector<Expression>& args,
           std::optional<Expression> repeat, const std::vector<Expression>& sinks,
           std::optional<AttributeSet> attrs = {}, std::optional<Expression> cond = {}, std::vector<Hook> hooks = {},
           Meta m = Meta())
         : NodeBase(nodes((id ? id : _uniquer.get("anon")), detail::adaptType(ctor.type(), repeat), ctor, repeat,
                          std::move(attrs), std::move(cond), args, sinks, std::move(hooks)),
                    std::move(m)),
+          _index(index),
           _is_anonynmous(! id.has_value()),
           _engine(e),
           _args_start(6),
           _args_end(_args_start + static_cast<int>(args.size())),
           _sinks_start(_args_end),
           _sinks_end(_sinks_start + static_cast<int>(sinks.size())) {}
 
-    Field(const std::optional<ID>& id, Item item, Engine e, const std::vector<Expression>& args,
+    Field(const std::optional<ID>& id, uint64_t index, Item item, Engine e, const std::vector<Expression>& args,
```

Comment:
It feels error-prone to have the caller put in charge of the indices; would be safer if the unit always did the numbering automatically behind the scenes. I see that it does that for `addItems()`, but not for the constructor. If I'm reading the patch right, that latter is because we'll be replacing fields later and need to set the index at that point, correct? Could we instead (1) leave the index out of the `Field` constructor and default it to zero, (2) change to Unit ctor to assign indices similar to `addItems()`, and (3) change `resolvefField` to preserve the original index for the new fields (through `new_field.setIndex(new_field, old_field.index())`?